hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/utils/TestSlidingWindow.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.common.utils;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+/**
+ * Tests for {@link SlidingWindow} class.
+ */
+public class TestSlidingWindow {
+
+  private SlidingWindow slidingWindow;
+
+  @BeforeEach
+  public void setup() {
+    slidingWindow = new SlidingWindow(3, Duration.ofSeconds(5));
+  }
+
+  @Test
+  public void testConstructorValidation() {
+    // Test invalid window size
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(0, Duration.ofMillis(100)));
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(-1, Duration.ofMillis(100)));
+
+    // Test invalid expiry duration
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(1, Duration.ofMillis(0)));
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(1, Duration.ofMillis(-1)));
+  }
+
+  @Test
+  public void testAdd() {
+    for (int i = 0; i < slidingWindow.getWindowSize(); i++) {
+      slidingWindow.add();
+      assertFalse(slidingWindow.isFull());
+    }
+
+    slidingWindow.add();
+    assertTrue(slidingWindow.isFull());
+  }
+
+  @Test
+  public void testEventExpiration() throws InterruptedException {
+    slidingWindow = new SlidingWindow(2, Duration.ofMillis(500));
+
+    // Add events to reach threshold
+    slidingWindow.add();
+    slidingWindow.add();
+    slidingWindow.add();
+    assertTrue(slidingWindow.isFull());
+
+    // Wait for events to expire
+    Thread.sleep(600);
```

Comment:
Please avoid tests based on sleep.  These introduce unnecessary delays and flakiness.  Let's use `Clock` in `SlidingWindow`, with a private implementation based on `Time.monotonicNow()` for real usage, and `TestClock` for the test.  Use `TestClock#fastForward` for controlling "time" without delay or flakiness.

https://github.com/apache/ozone/blob/fb75be7a8962b7ed57d75e0fed3fcd2befae2b6b/hadoop-hdds/test-utils/src/test/java/org/apache/ozone/test/TestClock.java#L26-L31